src/UnkObjects/unk8600F8.cpp

```diff
@@ -25,3 +26,201 @@ void unk_8600F8::loc_611DC0() {
     *x = this->unk14;
     this->unk14 = nullptr;
 }
+
+unk_612150 *unk_8600F8::sub_6124A0(const char *a) {
+    if (this->entries == NULL) {
+        return nullptr;
+    }
+
+    int index = get_entry_index(a);
+    
+    if (this->entries[index] == nullptr) {
+        return nullptr;
+    }
+
+    hash_map_entry *entry = this->entries[index];
+    
+    while (entry != nullptr) {
+        char *entryName = entry->name;
+
+        const char* a_p = a;
+        while (*a_p == *entryName) {
+
+            if (*entryName == '\0') {
+                return entry->value;
+            }
+            entryName++;
+            a_p++;
+        }
+        entry = entry->next;
+    }
+    return nullptr;
+}
+
+// mc2: 0x00611E70
+uint32_t unk_8600F8::get_entry_index(const char *a) {
+    if (this->entries == nullptr) {
+        this->entries = (hash_map_entry**)MC2_MALLOC(this->max_entries * 4);
+        std::memset(this->entries, 0, this->max_entries * 4);
+    }
+
+    uint32_t hash = 0;
+    for (; *a != '\0'; ++a) {
+        hash <<= 4;
+        hash += *a;
+        if (hash & 0xF0000000) {
+            hash ^= (hash & 0xF0000000) | ((hash & 0xF0000000) >> 24);
+        }
+    }
+    return hash % this->max_entries;
+}
+
+bool unk_8600F8::sub_612150(const char *a, unk_612150 *b) {
+    int index = get_entry_index(a);
+
+    if (this->entries[index] != NULL) {
+        hash_map_entry *inital_entry = this->entries[index];
+        hash_map_entry *entry = inital_entry;
+
+        while (entry != nullptr) {
+            char *entryName = entry->name;
+
+            const char *a_p = a;
+            while (*a_p == *entryName) {
+                if (*entryName == '\0') {
+                    return false;
+                }
+                entryName++;
+                a_p++;
+            }
+            entry = entry->next;
+        }
+    }
+
+    if (this->unk02 != 0) {
+        this->entries[index] = sub_611DF0(a, b, this->entries[index]);
+        this->num_entries++;
+        return true;
+    }
+
+    hash_map_entry *newEntry = (hash_map_entry*)MC2_MALLOC(sizeof(hash_map_entry));
+    if (newEntry != nullptr) {
+        newEntry->name = MC2_STRDUP(a);
+        newEntry->value = b;
+        newEntry->next = this->entries[index];
+    }
+
+    this->entries[index] = newEntry;
+    this->num_entries++;
+
+    if (this->num_entries <= this->max_entries) {
+        return true;
+    }
+    
+    if (glo_679778 == 0)
+        return true;
+
+    uint32_t esi_store = glo_8602C8;
+    glo_8602C8 = 5;
+    // Increase num entry size?
+    sub_6125A0(this->max_entries * 4);
+
+    glo_8602C8 = esi_store;
+
+    return true;
+}
+
+bool unk_8600F8::sub_611FE0(uint32_t *eax) {
+    if (this->entries == nullptr)
+        return false;
+
+    hash_map_entry *entry = sub_611F60(0);
+
+    if (entry == nullptr)
+        return false;
+
+    eax[0] = 0;
+    eax[1] = (uint32_t)entry->name;
+    eax[2] = (uint32_t)entry->value;
```

Comment:
One thing I **REALLY** want to avoid with this project is this sort of thing. In general, I want to have a rule that you can't have pointer-type to integer-type casts or vice versa (except in the address functions in Address.hpp). Since the original game was compiled from C++, it's a pretty safe bet that there were no such casts in the original game, so it's most likely that this sort of scenario is it filling up a struct similar to `hash_map_entry`. Keeping this rule will help prevent unpredictable pointer errors in the future.